tests/e2e/bin/includes.sh

```diff
@@ -0,0 +1,186 @@
+				read -t 1 reply </dev/tty
+
+				if [ -n "$reply" ]; then
+					break
+				fi
+			done
+		fi
+
+        # Default?
+        if [ -z "$reply" ]; then
+            reply=$default
+        fi
+
+        # Check if the reply is valid
+        case "$reply" in
+            Y*|y*) return 0 ;;
+            N*|n*) return 1 ;;
+        esac
+
+    done
+}
+
+##
+# Download from a remote source.
+#
+# Checks for the existence of curl and wget, then downloads the remote file using the first available option.
+#
+# @param {string} remote  The remote file to download.
+# @param {string} [local] Optional. The local filename to use. If it isn't passed, STDOUT is used.
+#
+# @return {bool} Whether the download succeeded or not.
+##
+download() {
+    if command_exists "curl"; then
+        curl -s -o "${2:--}" "$1"
+    elif command_exists "wget"; then
+		wget -nv -O "${2:--}" "$1"
+    fi
+}
+
+##
+# Add error message formatting to a string, and echo it.
+#
+# @param {string} message The string to add formatting to.
+##
+error_message() {
+	echo -en "\033[31mERROR\033[0m: $1"
+}
+
+##
+# Add warning message formatting to a string, and echo it.
+#
+# @param {string} message The string to add formatting to.
+##
+warning_message() {
+	echo -en "\033[33mWARNING\033[0m: $1"
+}
+
+##
+# Add status message formatting to a string, and echo it.
+#
+# @param {string} message The string to add formatting to.
+##
+status_message() {
+	echo -en "\033[32mSTATUS\033[0m: $1"
+}
+
+##
+# Add formatting to an action string.
+#
+# @param {string} message The string to add formatting to.
+##
+action_format() {
+	echo -en "\033[32m$1\033[0m"
+}
+
+##
+# Check if the command exists as some sort of executable.
+#
+# The executable form of the command could be an alias, function, builtin, executable file or shell keyword.
+#
+# @param {string} command The command to check.
+#
+# @return {bool} Whether the command exists or not.
+##
+command_exists() {
+	type -t "$1" >/dev/null 2>&1
+}
+
+# download and install ngrok
+install_ngrok() {
+	if $(command_exists "ngrok"); then
+			NGROK_CMD="ngrok"
+			return
+	fi
+
+	if [ -z "$CI" ]; then
+		echo -e $(error_message "Please install ngrok on your machine. Instructions: https://ngrok.com/download")
+		exit 1
+	fi
+
+	echo -e $(status_message "Installing ngrok in CI...")
+	curl -s https://bin.equinox.io/c/4VmDzA7iaHb/ngrok-stable-linux-amd64.zip > ngrok.zip
+	unzip ngrok.zip
+	NGROK_CMD="./ngrok"
+}
+
+start_ngrok() {
+	install_ngrok
+
+	echo -e $(status_message "Killing any rogue ngrok instances just in case...")
+	kill_ngrok
+
+	if [ ! -z "$NGROK_KEY" ]; then
+			$NGROK_CMD authtoken $NGROK_KEY
+	fi
+
+	$NGROK_CMD http -log=stdout 8889 > /dev/null &
+	sleep 3
+	WP_SITE_URL=$(get_ngrok_url)
+
+	if [ -z "$WP_SITE_URL" ]; then
+		echo -e $(error_message "WP_SITE_URL is not set after launching an ngrok")
+		exit 1
+	fi
+}
+
+get_ngrok_url() {
+	echo $(curl -s localhost:4040/api/tunnels/command_line | jq --raw-output .public_url)
```

Comment:
I am getting the following error on my machine:

`jq: command not found`

I can install it, but maybe we should take this into account somehow?

Reply:
Make sense. I'll take a look what's could be done here

Reply:
A `check_for_jq()` function would look something like `check_for_ngrok()`, perhaps.

Reply:
addressed in cf10a83